### PR TITLE
Temperature Pressure function fluid properties

### PR DESCRIPTION
**Pull request by CalfSteven**
Temperature pressure fluid function. A general fluid function that should not have restricting formulations. 
This source is made so it is allowed to take in both specific volume, internal energy formulations
and temperature, pressure formulations. The Density, Viscosity and thermal conductivity time derivative
are found by taking the gradient by using pressure and temperature as inputs.

### CODE REVIEW

**Comment by GiudGiud** [comment]
Should be fairly ready @joshuahansel if you want to have a look.
Unfortunately too much work to make this work for (v,e) but it s a good head start if someone wants to finish

File: modules/fluid_properties/doc/content/source/userobjects/TemperaturePressureFunctionFluidProperties.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+partial. Notable missing implementations are routines for entropy, the speed of sound, and some
+conversions between specific enthalpy and specific energy.
+
+!alert warning
+Due to the approximations made when computing the isobaric heat capacity from the constant
+isochoric heat capacity, this material should only be used for nearly-incompressible fluids.
+
+## Example Input File Syntax
+
+!listing modules/fluid_properties/test/tests/temperture_pressure_function/constant.i block=Modules
```
Comment (GiudGiud) [inline modules/fluid_properties/doc/content/source/userobjects/TemperaturePressureFunctionFluidProperties.md:23]:
```suggestion
!listing modules/fluid_properties/test/tests/temperature_pressure_function/constant.i block=FluidProperties
```

**Comment by joshuahansel** [comment]
@loganharbour I got 300+ emails (all in separate email threads) for "Your code requires style changes" for Job Precheck b266e1f. Any idea what's going on there?

**Comment by loganharbour** [comment]
> @loganharbour I got 300+ emails (all in separate email threads) for "Your code requires style changes" for Job Precheck [b266e1f](https://github.com/idaholab/moose/commit/b266e1fdb44b6aab41cb1a57bffb1882d4ae1ee7). Any idea what's going on there?

This happens every once in a while... we haven't been able to pin it down. But hey, at least it stopped?

Maybe this is our subtle way of making people scared of clang format and formatting before push... 😆

**Comment by joshuahansel** [comment]
> Maybe this is our subtle way of making people scared of clang format and formatting before push...

If that's the subtle way, I don't want to see the non-subtle way! 😄

File: modules/fluid_properties/src/materials/FluidPropertiesMaterialPT.C [hunk 2]
```diff
@@ -37,6 +42,9 @@ FluidPropertiesMaterialPT::FluidPropertiesMaterialPT(const InputParameters & par
     _s(declareProperty<Real>("s")),
     _c(declareProperty<Real>("c")),
```
Comment (joshuahansel) [inline modules/fluid_properties/src/materials/FluidPropertiesMaterialPT.C:43]: Only declare them if computed.

File: modules/fluid_properties/include/userobjects/TemperaturePressureFunctionFluidProperties.h [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+   * Temperature from specific volume and specific internal energy
+   *
+   * @param[in] v   specific volume (m$^3$/kg)
+   * @param[in] e   specific internal energy (J/kg)
+   * @return temperature (K)
+   */
+  virtual Real T_from_v_e(Real v, Real e) const override;
+
+  /**
+   * Pressure from pressure and density
```
Comment (joshuahansel) [inline modules/fluid_properties/include/userobjects/TemperaturePressureFunctionFluidProperties.h:44]:
```suggestion
   * Temperature from pressure and density
```

File: modules/fluid_properties/include/userobjects/TemperaturePressureFunctionFluidProperties.h [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+   *
+   * @param[in] p          pressure (Pa)
+   * @param[in] rho        density (kg/m$^3$)
+   * @return temperature (T)
+   */
+  virtual Real T_from_p_rho(Real p, Real rho) const;
+
+  /**
+   * Pressure from pressure and specific enthalpy
```
Comment (joshuahansel) [inline modules/fluid_properties/include/userobjects/TemperaturePressureFunctionFluidProperties.h:53]:
```suggestion
   * Temperature from pressure and specific enthalpy
```

File: modules/fluid_properties/src/userobjects/TemperaturePressureFunctionFluidProperties.C [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+
+InputParameters
+TemperaturePressureFunctionFluidProperties::validParams()
+{
+  InputParameters params = SinglePhaseFluidProperties::validParams();
+  params.addRequiredParam<FunctionName>("k", "Thermal conductivity function");
+  params.addRequiredParam<FunctionName>("rho", "Density function");
+  params.addRequiredParam<FunctionName>("mu", "Dynamic viscosity function");
+
+  params.addRequiredRangeCheckedParam<Real>("cv", "cv > 0", "Constant isochoric specific heat");
```
Comment (joshuahansel) [inline modules/fluid_properties/src/userobjects/TemperaturePressureFunctionFluidProperties.C:23]:
```suggestion
  params.addRequiredParam<FunctionName>("k", "Thermal conductivity function [W/(m-K)]");
  params.addRequiredParam<FunctionName>("rho", "Density function [kg/m^3]");
  params.addRequiredParam<FunctionName>("mu", "Dynamic viscosity function [Pa-s]");

  params.addRequiredRangeCheckedParam<Real>("cv", "cv > 0", "Constant isochoric specific heat [J/(kg-K)]");
```

File: modules/fluid_properties/test/tests/temperature_pressure_function/exact.i [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    type = DifferencePostprocessor
+    value1 = cv_exact
+    value2 = cv_avg
+  []
+[]
+
+[Outputs]
+  # Note that diffs wont be settled until timestep 2 because of order of execution
+  csv = true
+  exodus = true
```
Comment (joshuahansel) [inline modules/fluid_properties/test/tests/temperature_pressure_function/exact.i:180]:
```suggestion
```

File: modules/fluid_properties/test/tests/temperature_pressure_function/exact.i [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+  []
+  [cv_diff]
+    type = DifferencePostprocessor
+    value1 = cv_exact
+    value2 = cv_avg
+  []
+[]
+
+[Outputs]
+  # Note that diffs wont be settled until timestep 2 because of order of execution
```
Comment (joshuahansel) [inline modules/fluid_properties/test/tests/temperature_pressure_function/exact.i:178]: Doesn't matter much, but this surprises me. This seems like something the dependence resolution should be able to handle just fine.

**Review by joshuahansel — CHANGES_REQUESTED** [review]

File: modules/fluid_properties/doc/content/source/userobjects/TemperaturePressureFunctionFluidProperties.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
+# TemperaturePressureFunctionFluidProperties
+
+!syntax description /FluidProperties/TemperaturePressureFunctionFluidProperties
+
+The derivatives of the fluid properties are obtained using the `Function`(s) gradient components
+and the appropriate derivative chaining for derived properties.
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/fluid_properties/doc/content/source/userobjects/TemperaturePressureFunctionFluidProperties.md:6]: added
You are very right, this was confusing

File: modules/fluid_properties/doc/content/source/userobjects/TemperaturePressureFunctionFluidProperties.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+partial. Notable missing implementations are routines for entropy, the speed of sound, and some
+conversions between specific enthalpy and specific energy.
+
+!alert warning
+Due to the approximations made when computing the isobaric heat capacity from the constant
+isochoric heat capacity, this material should only be used for nearly-incompressible fluids.
+
+## Example Input File Syntax
+
+!listing modules/fluid_properties/test/tests/temperature_pressure_function/exact.i block=FluidProperties
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/fluid_properties/doc/content/source/userobjects/TemperaturePressureFunctionFluidProperties.md:23]: added some text

File: modules/fluid_properties/src/userobjects/TemperaturePressureFunctionFluidProperties.C [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+  rho_from_p_T(pressure, temperature, rho, drho_dp, drho_dT);
+  h = e + pressure / rho;
+  dh_dp = de_dp + 1 / rho - pressure / rho / rho * drho_dp;
+  dh_dT = de_dT - pressure * drho_dT / rho / rho;
+}
+
+Real
+TemperaturePressureFunctionFluidProperties::e_from_p_T(Real /* pressure */, Real temperature) const
+{
+  return _cv * temperature;
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/fluid_properties/src/userobjects/TemperaturePressureFunctionFluidProperties.C:239]: done

File: modules/fluid_properties/test/tests/temperature_pressure_function/exact.i (hunk 7 above)
Comment (GiudGiud) [reply modules/fluid_properties/test/tests/temperature_pressure_function/exact.i:178]: I played around with preaux and postaux but could not get resolution.
There's no cross resolution between auxes and UOs so it s not too surprising

File: modules/fluid_properties/src/materials/FluidPropertiesMaterialPT.C (hunk 2 above)
Comment (GiudGiud) [reply modules/fluid_properties/src/materials/FluidPropertiesMaterialPT.C]: done

**Review by GiudGiud — COMMENTED** [review]
Review should be addressed

File: modules/fluid_properties/test/tests/temperature_pressure_function/exact.i (hunk 7 above)
Comment (joshuahansel) [reply modules/fluid_properties/test/tests/temperature_pressure_function/exact.i:178]: Ah, I didn't notice the aux dependence.

**Comment by joshuahansel** [comment]
Looks good, just need tests to pass.

**Comment by GiudGiud** [comment]
ah civet is hosed. invalidated the bad ones

**Comment by GiudGiud** [comment]
so green now

**Comment by joshuahansel** [comment]
Looks like the following might need some line coverage:

- `rho_from_p_T(AD, AD, AD, AD, AD)`
- `v_from_p_T(Real, Real)`
- `v_from_p_T(Real, Real, Real, Real, Real)`

I don't care too much about the first one because I don't think many people use that interface.

**Comment by GiudGiud** [comment]
Added for all 3

**Review by joshuahansel — APPROVED** [review]

**Comment by GiudGiud** [comment]
Thanks for the prompt review!

**Pull request merged**